libobs/obs-cocoa.m

```diff
@@ -889,8 +894,48 @@ bool obs_hotkeys_platform_is_pressed(obs_hotkeys_platform_t *plat,
 	return plat->is_key_down[key];
 }
 
+static void mach_set_thread_realtime(uint64_t period, double busy,
+				     double busy_limit, bool preemptible)
+{
+	mach_timebase_info_data_t timebase_info;
+	mach_timebase_info(&timebase_info);
+
+	double realtime_period =
+		((double)timebase_info.denom / (double)timebase_info.numer) *
+		period;
+
+	thread_time_constraint_policy_data_t policy;
+	policy.period = (uint32_t)realtime_period;
+	policy.computation = (uint32_t)(realtime_period * busy);
+	policy.constraint = (uint32_t)(realtime_period * busy_limit);
+
+	policy.preemptible = preemptible;
+
+	int kr = thread_policy_set(pthread_mach_thread_np(pthread_self()),
+				   THREAD_TIME_CONSTRAINT_POLICY,
+				   (thread_policy_t)&policy,
+				   THREAD_TIME_CONSTRAINT_POLICY_COUNT);
+	if (kr != KERN_SUCCESS) {
+		blog(LOG_ERROR,
+		     "error setting rt thread, thread_policy_set: %s",
+		     mach_error_type(kr));
+	}
+
+	pthread_set_qos_class_self_np(QOS_CLASS_USER_INTERACTIVE, 0);
```

Comment:
I don't know if we want pthread_set_qos_class_self_np. If it's not a huge win, I'd rather not cross the streams and make things more complicated.

Reply:
It does not appear to alter the behavior of manually specifying constraints

Reply:
My impression of the Apple video is that it's using pthread because it doesn't want games to use realtime priority. An argument could be made that the OBS graphics thread hitting its deadline is more critical than a game starting its render thread from a consistent point.

We should probably use one or the other, either the mach APIs for realtime priority, or pthread APIs for high priority, and I wouldn't mind staying with mach. At least, that seems cleaner to me from this documentation:
https://developer.apple.com/library/archive/documentation/Darwin/Conceptual/KernelProgramming/scheduler/scheduler.html

Reply:
Sorry, by my statement I was in support of removing as it appears to add nothing. (Just for clarification)

Reply:
No worries. I hadn't even seen your reply as I was writing that wall of text. :P

Reply:
Yeah I think James is right that QoS has probably no bearing on realtime threads, though it sounds like we should not use them in the first place (if games are discouraged from it).

Note that all these "old" documents from the Developer archive do not take the changes/differences of the new Apple Silicon architecture into account, so things might actually not fully apply anymore.

Reply:
Though, from my light testing, utilizing only QoS doesn't quite get us into the accuracy of timers we'd prefer

Reply:
Did you try with `SCHED_RR` and QoS 45 as the video suggested? Otherwise I'm happy to defer to James' judgement.

Reply:
Then I say Mach it up!

Reply:
Sched RR disables qos from my understanding, I'm afk but I'll double check

Reply:
Yeah it's then called "priority" it seems:

<img width="973" alt="Bildschirm­foto 2022-12-21 um 18 41 30" src="https://user-images.githubusercontent.com/9436503/208972231-d0c470bb-1837-4225-9a7f-26269574e669.png">

Reply:
info: obs_graphics_thread(16.6667 ms): min=0.22 ms, median=16.666 ms, max=73.189 ms, 63.9489% within ±2% of 16.667 ms (18.4185% lower, 17.6326% higher)

Sched RR 45

Reply:
Reverted QoS addition and switched to .75/.85